### PR TITLE
[JENKINS-56927] Implementation of credentials for private SSH keys used by EC2 plugin

### PR DESCRIPTION
**Pull request by djesionek**
This should fix the issue described in [JENKINS-56927](https://issues.jenkins-ci.org/browse/JENKINS-56927).

I tried to keep as much of the credential logic itself contained inside the _EC2Cloud_ class but still in a way that the private key itself is not being stored longer than necessary outside of the credential API.

This implementation also takes care of migrating the old plaintext ssh configuration to the credentials API. This was checked in a test environment and also the unit tests themselves are not modified more than necessary and some of them still relied on the old storage method and a migration is performed in each one. 
In my tests the privateKey was still included in plaintext in the CasC yaml when exporting, so I managed to find a quite hacky solution for that by returning _null_ in the getter and creating a static method resolving the credential of a _EC2Cloud_ object to a _EC2PrivateKey_ which is now used instead in all places. If there is a better way for that I have not found it yet...

All of this introduced one new dependency: _org.jenkins-ci.plugins:ssh-credentials:1.14_
As the _aws-credentials_ plugin has a different dependency to the _credentials_ package it also was fixed to be min version 2.1.17 to prevent dependency errors. I have not checked how low we could go here but I'm also not sure how important this might be here.

Eucalyptus implementations also were updated accordingly with this but could not be tested explicitly due to lack of a test environment on my side.

### CODE REVIEW

**Comment by djesionek** [comment]
Hi @jvz and @res0nance ,
should I have assigned anyone from you maintainers as reviewer for this PR to be picked up or is something else missing?

**Comment by jvz** [comment]
Ooh, this looks nice! I've added some people as reviewers to look through this.

**Comment by Wadeck** [comment]
_(I will not have the time to review it, sorry)_

File: pom.xml
Comment (jvz) [inline pom.xml:107]:
```suggestion
            <version>2.3.5</version>
```

This is fairly old otherwise.

File: pom.xml
Comment (jvz) [inline pom.xml:117]:
```suggestion
            <version>1.18.1</version>
```

Same.

File: src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java:200]: Needs a permission check for `Node/Configure` or admin.

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:988]: Now that this method is looking up system credentials, it should have a permission check added.

File: src/main/java/hudson/plugins/ec2/Eucalyptus.java
Comment (jvz) [inline src/main/java/hudson/plugins/ec2/Eucalyptus.java:90]: This also needs a permission check.

File: src/main/resources/hudson/plugins/ec2/AmazonEC2Cloud/config-entries.jelly
Comment (jvz) [inline src/main/resources/hudson/plugins/ec2/AmazonEC2Cloud/config-entries.jelly:40]:
```suggestion
    <c:select includeUser="false"/>
```

User-scoped credentials won't be relevant here as there's no build associated.

File: src/main/resources/hudson/plugins/ec2/Eucalyptus/config-entries.jelly
Comment (jvz) [inline src/main/resources/hudson/plugins/ec2/Eucalyptus/config-entries.jelly:35]:
```suggestion
    <c:select includeUser="false"/>
```

Same here.

**Comment by djesionek** [comment]
@jvz how do you handle changing the dependency versions generally? After applying your changes and setting newer dep versions the enforcer requires me to also change trilead-api 1.0.3->1.0.5 and structs 1.17->1.20. The newer version of ssh-credentials you suggested also requires at least jenkins at version 2.190.1. 
Are there any considerations to make regarding backwards compatibility of any sorts or should they be just the newest version at the time of implementation of features?

**Comment by jvz** [comment]
Jenkins 2.190.1 is fairly old as it is, so that's not a bad baseline now. The current LTS is in 2.222.x, so that's a few LTS releases behind already. I'll defer that to one of the other maintainers, though.

**Comment by djesionek** [comment]
> Jenkins 2.190.1 is fairly old as it is, so that's not a bad baseline now. The current LTS is in 2.222.x, so that's a few LTS releases behind already. I'll defer that to one of the other maintainers, though.

I'm still trying to find the right combination of dependency versions as bumping the jenkins version results only in more dependency issues that I can not interpret at this point. So suggestions are much appreciated

**Comment by jvz** [comment]
Perhaps we need to adopt a Jenkins [bom](https://github.com/jenkinsci/bom/)?

**Comment by djesionek** [comment]
> Perhaps we need to adopt a Jenkins [bom](https://github.com/jenkinsci/bom/)?
> 
> Edit: which has apparently moved into [the core repo](https://github.com/jenkinsci/jenkins/blob/master/bom/pom.xml).

That might depend on whether we want to or will be able stay below Jenkins 2.195

**Comment by djesionek** [comment]
@jvz you were right, the bom is apparently one solution. Maybe this helps preventing such hassle in future. Is this something I should include here in this PR probably with some cleanup (of the plugin versions no longer necessary in the pom) or leave the cleanup for another task?

**Comment by jvz** [comment]
You can do the dependency update and bom in a separate PR.

**Comment by djesionek** [comment]
> You can do the dependency update and bom in a separate PR.

Does it make sense then to revert the dependency changes here so the PR can be actually built and then reviewed independently to the dependency topic?

**Comment by jvz** [comment]
> Does it make sense then to revert the dependency changes here so the PR can be actually built and then reviewed independently to the dependency topic?

Yeah, go ahead. Thanks!

File: pom.xml
Comment (res0nance) [inline pom.xml:149]: That comment should be dropped

File: pom.xml
Comment (res0nance) [inline pom.xml:267]: wasn't the whole point of bom to stop having to specify specific plugin versions?

File: src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/AmazonEC2Cloud.java:52]: No star imports please

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:39]: Same here

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:52]: Here as well

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:147]: the `privateKey` variable should not be removed it should be marked as transient

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:307]:
```suggestion
    @CheckForNull
    public String getSshKeysCredentialsId() {
```

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:943]:
```suggestion
    @CheckForNull
    public static EC2PrivateKey resolvePrivateKey(EC2Cloud cloud){
```

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:364]: This method can return null so this can result in an NPE

File: src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java
Comment (res0nance) [inline src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java:44]: Why do we need jenkinsrule?

File: src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java
Comment (res0nance) [inline src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java:50]: All the tests fail without this?

File: pom.xml
Comment (jvz) [reply pom.xml:267]: Yes; the version overrides specified in pom.xml can be removed with this bom applied.

File: src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java
Comment (jvz) [reply src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java:50]: Probably an issue in Java 9+? I read about an incompatibility with the module path and how JCA is designed, but I can't find an official bug about it.

File: src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java
Comment (djesionek) [reply src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java:44]: _AmazonEC2Cloud_ now relies on a call to SystemCredentialsProvider.getInstance() which is not available without a jenkinsrule. Is there another way to solve this?

File: src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java
Comment (djesionek) [reply src/test/java/hudson/plugins/ec2/EC2PrivateKeyTest.java]: For some reason this is not possible anymore on my current machine. I removed it for now.

File: src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java
Comment (res0nance) [reply src/test/java/hudson/plugins/ec2/AmazonEC2CloudUnitTest.java:44]: AFAIK, no. If it is needed thats fine then

File: pom.xml
Comment (djesionek) [reply pom.xml:267]: Removing the versions from all plugin deps does not work for me. I also couldn't find a reference that this should work in neither bom code ( core-bom and plugin-bom )

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:164]: @res0nance is there any reason to leave it marked as transient if it is not being used at all?
As far as I understand it it will still be passed through the constructor after a plugin update from a version where it was used and will be migrated to a credential. So this would be left as an unused variable.

**Comment by jvz** [comment]
I don't maintain this plugin anymore, and I can't see what changed since my last review. Do you have a diff?

**Comment by djesionek** [comment]
> I don't maintain this plugin anymore, and I can't see what changed since my last review. Do you have a diff?

The first commit after you requested changes was ab02e81 so you should be able to diff locally or on github from this commit on.

**Comment by jvz** [comment]
Thanks for the tip! That made things easier to re-review.

**Comment by djesionek** [comment]
@res0nance is there anything I might have missed that is preventing further reviews?

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:164]: IIRC it wont call the constructor, it might cause deserialization issues if the field no longer exists

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:186]: This should probably be in `readResolve()`

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:1051]: I'm not 100% certain but it might need to `@RequirePOST`

File: src/test/java/hudson/plugins/ec2/SlaveTemplateTest.java
Comment (res0nance) [inline src/test/java/hudson/plugins/ec2/SlaveTemplateTest.java:90]: Empty?

File: src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java:172]: If I have to decrypt the password but there is no private key is there any point to continuing?

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:1005]:
```suggestion
    @CheckForNull
    private static BasicSSHUserPrivateKey getSshCredential(String id){
```

File: src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java:172]: I was not sure what other way to go, but continuing with the empty password will most definitely fail which should happen anyways in the case of a missing private key. Any suggestion for better error handling here?

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1051]: No other doCheckX method has this annotation in place. How to find out?

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:186]: According to the comment at the readResolve() method call I interpret it as method caring about setting the template parents while this part checks if a migration to credentials is necessary here which in my opinion would not be part of this topic. Am I missing something here?

File: src/test/java/hudson/plugins/ec2/SlaveTemplateTest.java
Comment (djesionek) [reply src/test/java/hudson/plugins/ec2/SlaveTemplateTest.java:90]: Forget to clean up, will be deleted.

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1011]: Added it

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:167]: My observations in debugging showed that the constructor is called here and the migration seemed to also work but should maybe be checked to be removed in future. I rewrote it to similar to the handling of the other deprecated and transient attributes now.

**Comment by djesionek** [comment]
The build fails because it uses version 2.164.1 of Jenkins for some reason now. This changed in the last days. Before versions above the required version 2.190.3 according to my updated pom.xml were used.
@res0nance will your PR #465 fix this issue or do I need to change something else here?

**Comment by djesionek** [comment]
@res0nance @jvz @julienduchesne @thoulen what am I missing that prevents further reviews to happen? Some of my comments are still open for your feedback.
I would be happy to fix everything necessary to get this through before bigger changes in the master branch lead to further merge-conflicts.

**Comment by jvz** [comment]
I've already approved this.

File: src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java
Comment (res0nance) [reply src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java:172]: We should probably log it and throw an exception. I believe it will terminate the instance in this case

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (res0nance) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: @jvz @MRamonLeon do either of you know if this is required?

File: src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java:172]: A few lines above the missing passwordData is handled by calling `continue` in the whole connection loop with the expectation that the passwordData will be added between the loops. 
Do you think it would lead to other issues handling a null privateKey exactly like that, expecting that the logs are read an the user configured the right credential?

```suggestion
                        if (ec2PrivateKey == null){
                            logger.println("Waiting for privateKey to be available. Consider checking the credentials in the cloud configuration. Sleeping 10s. ");
                            Thread.sleep(sleepBetweenAttempts);
                            continue;
                        }
                        String password = ec2PrivateKey.decryptWindowsPassword(passwordData);
```

Seems for me like a legit way to handle this as nothing gets terminated but a correct configuration gets enforced.

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (markyjackson-taulia) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: There was a ping on twitter regarding this Pr and jira issue this is addressing so I thought I would jump in to help. A @RequirePOST is required when doing `doCheckX`
@djesionek

Comment (MRamonLeon) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: I would say no, as long as the method is not modifying any state. Maybe @daniel-beck can confirm.

File: src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java
Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/win/EC2WindowsLauncher.java]: I now made this change as it makes sense to me like that. Thanks for pointing that out

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (daniel-beck) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: https://www.jenkins.io/doc/developer/security/form-validation/ explains it.

At a glance, depends on exactly what `getSshCredential` does. If there are potentially network lookups for credentials involved, it's safer to `@RequirePOST` (or `@POST`).

Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: It only calls the `CredentialsProvider`, so `@RequirePOST` is not necessary in that case?

Comment (daniel-beck) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: Unsure how expensive network lookups can become with external credentials providers.

I mean, when in doubt add it; it's a reasonable default.

Comment (djesionek) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1057]: Makes sense, didn't thought about external credentials providers at all. So I will add it. Thanks :+1:

File: pom.xml
Comment (djesionek) [reply pom.xml:264]: @res0nance is this topic holding back further reviews and approvals?

**Comment by froblesmartin** [comment]
What is missing on this PR to be merged? @djesionek @markyjackson-taulia

**Comment by djesionek** [comment]
AFAIK I adressed all the requested changes from the reviewers that reviewed this PR. I'm just waiting for either feedback or a merge.
Maybe @res0nance can tell what is missing.

**Comment by froblesmartin** [comment]
> AFAIK I adressed all the requested changes from the reviewers that reviewed this PR. I'm just waiting for either feedback or a merge.
> Maybe @res0nance can tell what is missing.

Well, at least what you can see below, there is a conflict on the EC2Cloud.java file

**Comment by djesionek** [comment]
Thanks for the info, that one seems to be new. I won't be able to resolve it until Sunday probably.

**Comment by froblesmartin** [comment]
@djesionek FYI in https://issues.jenkins-ci.org/browse/SECURITY-1883 @daniel-beck asked @jetersen to have a look into this.

**Comment by djesionek** [comment]
@froblesmartin thanks for the info somehow my jenkins community account got lost that's probably why I missed that.

**Comment by jetersen** [comment]
not sure what I can assist with.

**Comment by djesionek** [comment]
And I'm not sure what is missing for getting this merged other than maybe further reviews when necessary. Maybe @daniel-beck knows more?

**Comment by jvz** [comment]
This plugin needs active maintainers. I've re-added the `adopt-this-plugin` label as this hasn't received new maintainers since I thought I last added that.

**Comment by froblesmartin** [comment]
> This plugin needs active maintainers. I've re-added the `adopt-this-plugin` label as this hasn't received new maintainers since I thought I last added that.

@jvz I have been thinking since a while about maintaining a plugin to start contributing. This could be a perfect moment 😋 .

Following https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/, I am going to send an email to the DL requesting access 👍

**Review by res0nance — APPROVED** [review]

**Comment by djesionek** [comment]
Thanks, hope it gets released soon 😄

**Comment by batmat** [comment]
@res0nance @MRamonLeon @fcojfernandez @froblesmartin could you please consider doing a release for this? Thanks!

**Comment by MRamonLeon** [comment]
Released as https://github.com/jenkinsci/ec2-plugin/releases/tag/ec2-1.52

**Comment by varyvol** [comment]
How was migration tested here? I had one EC2 cloud configured with 1.50.3 and when updated to 1.52, the cloud is there but no credentials are in place for the private key. Also an error is displayed because the "doCheckSshKeysCredentialsId" requires POST and the validation performed when the credential is selected is not using that.

**Comment by MRamonLeon** [comment]
^ @res0nance

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (daniel-beck) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:175]: A migration call in the constructor makes no sense. This should be in `readResolve`. This was probably never tested.

**Comment by djesionek** [comment]
I made a PR with a fix for that and retested the migration scenario there manually. Maybe it would be good to create an automated test for that but at this point I have no clue how this could be achieved 😄 
Sorry that I missed this!

**Comment by erik-hakansson-wcar** [comment]
This broke our setup as the old key wasn't migrated.

**Comment by varyvol** [comment]
ec2-1.53 fixes it.

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java
Comment (jglick) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java:1027]: [As noted in Jira](https://issues.jenkins.io/browse/JENKINS-56927?focusedId=398071&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-398071) this is incorrect: should be using the interface type, `SSHUserPrivateKey`.

Comment (jglick) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1027]: I see #602 changing other things. I think you meant #520.

Comment (res0nance) [reply src/main/java/hudson/plugins/ec2/EC2Cloud.java:1027]: Oh yes you're right!

**Pull request merged**
